### ISSUE TITLE
[Question] Suitability for mesh similarity measurement

### ISSUE THREAD

**Issue by Maghoumi** [hexygen/shed#1]
First, kudos to you for sharing the well-documented code 👍 
I have a few questions and I'd really appreciate your insight:

1) For a task in our project, we need to be able to compare an extracted mesh of an object to the ground truth mesh of that same object. In other words, given a mesh and a ground truth, we are interested in determining how similar they are. Is SHED a suitable similarity metric for this purpose?

2) I've been able to successfully run SHED on a few of our data samples. Since our task is not segmentation, I went ahead and provided the same segmentation label to all the faces of both meshes (so forced SHED to treat the whole object as a single segment). I was able obtain the results, and as you would expect SHED reported small error values for shapes that looked similar. If the answer to Q1 is yes, is this the correct approach of using SHED for our purpose?

3) Based on the SHED paper, I can see that the weighted sum of various terms are used in computing the SHED value. Is it correct to assume that the SHED algorithm outputs error values in the range [0, 1] (0 being the best value, 1 being the worst value)? Or does SHED output values in other numerical range?

I wanted to email you, but when I found the repository, I thought my question may help other people if they decide to use SHED. Thanks in advance! :-)

**Comment by hexygen** [comment]
Hi Maghoumi,
Thanks for the interest in our project!

1. You did not give a lot of information here but it sounds that this is not the ideal similarity metric for your purpose. SHED is not designed to capture differences in small details such as sampling differences, vertex perturbations, noise or for example isometric deformation (i.e. changing the pose of the model). In fact, it is designed to be robust to these changes, to be able to identify similar shapes which have different low-level representations. For example, if a shape is re-sampled or re-scanned, the SHED between the shapes *should* be negligible. For shapes with the same structure and overall similar geometry, there is no guarantee that the more accurate representation (in terms of noise, etc) will have a lower SHED.

2. The answer to this is **NO**. This is not the correct way to use SHED. SHED Depends on the structure of the shape, which is derived from the provided segmentation. We recommend using the [Approximate Convexity Analysis](http://www.cs.tau.ac.il/~noafish/wcseg/) segmentation if possible. Without the segmentation, SHED has no structural information of the shape and is reduced to report the geometric differences between the single segment you provided for each shape. This geometric difference is computed using Shape Distributions (see [here](http://graphics.stanford.edu/courses/cs468-08-fall/pdf/osada.pdf)), namely D1 and D2 descriptors. D1 measures a histogram of the distances between sampled points on the shape and the center of mass. D2 measures a histogram of the distances between pairs of sampled points on the shape. This is the result you got - these descriptors will work when comparing very different models, but they are not super sensitive (they work well for SHED because they are applied to small nearly convex segments). 

3. No, you cannot assume SHED values are capped at [0, 1]. The theoretic minimum is 0 but since the descriptors mentioned above are randomly sampled the distance is not likely to be 0 even when comparing a shape to itself (however, it will be very small). Since shapes can be different in a lot of ways there is no bound to how two shapes may differ and SHED can be very large in extreme cases.

To summarize, if you want to use SHED please make sure the shapes you are comparing are segmented in a way which reveals the shape's structure and is robust/predictable (i.e. similar shapes have similar segmentations). If I understand your purpose correctly, you might want to consider a more local approach like [MeshGit](http://dl.acm.org/citation.cfm?id=2461942).

Best of luck with your project!
